File: python-sdk/ade_sidechain/transaction.py
"""Transaction builder and utilities"""

import hashlib
from typing import List, Optional
from dataclasses import dataclass


@dataclass
class MessageHeader:
    """Transaction message header"""
    num_required_signatures: int
    num_readonly_signed_accounts: int
    num_readonly_unsigned_accounts: int


@dataclass
class CompiledInstruction:
    """Compiled instruction"""
    program_id_index: int
    accounts: List[int]
    data: bytes


@dataclass
class Message:
    """Transaction message"""
    header: MessageHeader
    account_keys: List[bytes]
    recent_blockhash: bytes
    instructions: List[CompiledInstruction]


class Transaction:
    """Transaction object"""

    def __init__(self, message: Message):
        self.message = message
        self.signatures: List[bytes] = []

    def hash(self) -> bytes:
        """Compute transaction hash"""
        serialized = self.serialize()
        return hashlib.sha256(serialized).digest()

    def serialize(self) -> bytes:
        """Serialize transaction to bytes"""
        parts = []
        
        # Serialize signatures count
        parts.append(bytes([len(self.signatures)]))
        
        # Serialize signatures
        for sig in self.signatures:
            parts.append(sig)
        
        # Serialize message
        parts.append(self._serialize_message())
        
        return b''.join(parts)

    def _serialize_message(self) -> bytes:
        """Serialize message to bytes"""
        parts = []
        
        # Header
        parts.append(bytes([
            self.message.header.num_required_signatures,
            self.message.header.num_readonly_signed_accounts,
            self.message.header.num_readonly_unsigned_accounts,
        ]))
        
        # Account keys
        parts.append(self._encode_length(len(self.message.account_keys)))
        for key in self.message.account_keys:
            parts.append(key)
        
        # Recent blockhash
        parts.append(self.message.recent_blockhash)
        
        # Instructions
        parts.append(self._encode_length(len(self.message.instructions)))
        for instruction in self.message.instructions:
            parts.append(bytes([instruction.program_id_index]))
            parts.append(self._encode_length(len(instruction.accounts)))
            parts.append(bytes(instruction.accounts))
            parts.append(self._encode_length(len(instruction.data)))
            parts.append(instruction.data)
        
        return b''.join(parts)

    @staticmethod
    def _encode_length(length: int) -> bytes:
        """Encode length as compact-u16"""
        result = []
        while length > 0x7f:
            result.append((length & 0x7f) | 0x80)
            length >>= 7
        result.append(length)
        return bytes(result)


class TransactionBuilder:
    """Builder for constructing transactions"""

    def __init__(self):
        self.instructions: List[CompiledInstruction] = []
        self.account_keys: List[bytes] = []
        self.recent_blockhash: Optional[bytes] = None

    def add_instruction(self, instruction: CompiledInstruction) -> 'TransactionBuilder':
        """Add an instruction"""
        self.instructions.append(instruction)
        return self

    def add_account_key(self, key: bytes) -> 'TransactionBuilder':
        """Add an account key"""
        self.account_keys.append(key)
        return self

    def set_recent_blockhash(self, blockhash: bytes) -> 'TransactionBuilder':
        """Set recent blockhash"""
        self.recent_blockhash = blockhash
        return self

    def build(self) -> Transaction:
        """Build the transaction"""
        if self.recent_blockhash is None:
            raise ValueError("Recent blockhash not set")

        header = MessageHeader(
            num_required_signatures=len(self.account_keys),
            num_readonly_signed_accounts=0,
            num_readonly_unsigned_accounts=0,
        )

        message = Message(
            header=header,
            account_keys=self.account_keys,
            recent_blockhash=self.recent_blockhash,
            instructions=self.instructions,
        )

        return Transaction(message)

File: src/transaction.ts
import { Keypair, PublicKey } from '@solana/web3.js';
import * as crypto from 'crypto';

export interface MessageHeader {
  numRequiredSignatures: number;
  numReadonlySignedAccounts: number;
  numReadonlyUnsignedAccounts: number;
}

export interface Message {
  header: MessageHeader;
  accountKeys: Uint8Array[];
  recentBlockhash: Uint8Array;
  instructions: CompiledInstruction[];
}

export interface CompiledInstruction {
  programIdIndex: number;
  accounts: number[];
  data: Uint8Array;
}

export class Transaction {
  public signatures: Uint8Array[] = [];
  public message: Message;

  constructor(message: Message) {
    this.message = message;
  }

  sign(signers: Keypair[]): void {
    const messageBytes = this.serializeMessage();
    
    this.signatures = signers.map(signer => {
      return signer.sign(messageBytes).signature;
    });
  }

  serialize(): Uint8Array {
    const messageBytes = this.serializeMessage();
    const signaturesLength = this.signatures.length;
    
    let totalLength = 1 + (signaturesLength * 64) + messageBytes.length;
    const buffer = new Uint8Array(totalLength);
    
    let offset = 0;
    buffer[offset++] = signaturesLength;
    
    for (const signature of this.signatures) {
      buffer.set(signature, offset);
      offset += 64;
    }
    
    buffer.set(messageBytes, offset);
    
    return buffer;
  }

  private serializeMessage(): Uint8Array {
    const buffers: Uint8Array[] = [];
    
    buffers.push(new Uint8Array([
      this.message.header.numRequiredSignatures,
      this.message.header.numReadonlySignedAccounts,
      this.message.header.numReadonlyUnsignedAccounts,
    ]));
    
    buffers.push(this.encodeLength(this.message.accountKeys.length));
    for (const key of this.message.accountKeys) {
      buffers.push(key);
    }
    
    buffers.push(this.message.recentBlockhash);
    
    buffers.push(this.encodeLength(this.message.instructions.length));
    for (const instruction of this.message.instructions) {
      buffers.push(new Uint8Array([instruction.programIdIndex]));
      buffers.push(this.encodeLength(instruction.accounts.length));
      buffers.push(new Uint8Array(instruction.accounts));
      buffers.push(this.encodeLength(instruction.data.length));
      buffers.push(instruction.data);
    }
    
    const totalLength = buffers.reduce((acc, buf) => acc + buf.length, 0);
    const result = new Uint8Array(totalLength);
    let offset = 0;
    
    for (const buffer of buffers) {
      result.set(buffer, offset);
      offset += buffer.length;
    }
    
    return result;
  }

  private encodeLength(length: number): Uint8Array {
    const bytes: number[] = [];
    let len = length;
    
    while (len > 0x7f) {
      bytes.push((len & 0x7f) | 0x80);
      len >>= 7;
    }
    bytes.push(len);
    
    return new Uint8Array(bytes);
  }

  hash(): Uint8Array {
    const serialized = this.serialize();
    return crypto.createHash('sha256').update(serialized).digest();
  }
}

export class TransactionBuilder {
  private instructions: CompiledInstruction[] = [];
  private signers: Keypair[] = [];
  private recentBlockhash?: Uint8Array;

  addInstruction(instruction: CompiledInstruction): this {
    this.instructions.push(instruction);
    return this;
  }

  addSigner(signer: Keypair): this {
    this.signers.push(signer);
    return this;
  }

  setRecentBlockhash(blockhash: Uint8Array): this {
    this.recentBlockhash = blockhash;
    return this;
  }

  build(): Transaction {
    if (!this.recentBlockhash) {
      throw new Error('Recent blockhash not set');
    }

    const accountKeys = this.signers.map(s => s.publicKey.toBytes());

    const message: Message = {
      header: {
        numRequiredSignatures: this.signers.length,
        numReadonlySignedAccounts: 0,
        numReadonlyUnsignedAccounts: 0,
      },
      accountKeys,
      recentBlockhash: this.recentBlockhash,
      instructions: this.instructions,
    };

    const transaction = new Transaction(message);
    transaction.sign(this.signers);

    return transaction;
  }
}






